kv/kvserver: TestMergeReplicatesLocks failed

kv/kvserver.TestMergeReplicatesLocks [failed](https://tanzanite.cluster.engflow.com/invocations/default/1ee40d3b-a12c-462c-9c19-191f01298b26?testReportRun=2&testReportShard=41&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-25.2.0-rc @ [545da2936c70fdecb9bbbd652aa4bd90b8d26dad](https://github.com/cockroachdb/cockroach/commits/545da2936c70fdecb9bbbd652aa4bd90b8d26dad):


```
=== RUN   TestMergeReplicatesLocks
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestMergeReplicatesLocks2241698621
    test_log_scope.go:76: use -show-logs to present logs inline
    client_replica_test.go:6208: -- test log scope end --
test logs left over in: outputs.zip/logTestMergeReplicatesLocks2241698621
--- FAIL: TestMergeReplicatesLocks (74.13s)
=== RUN   TestMergeReplicatesLocks/rhs-lock
    client_replica_test.go:6159: txn1: lock acquired, waiting for txn2 cancellation
    client_replica_test.go:6171: txn2: tnx1 lock acquisition observed, starting txn
    client_replica_test.go:6192: merging ranges
    client_replica_test.go:6196: cancelling txn2
    client_replica_test.go:6161: txn1: done
    client_replica_test.go:6204: consistency failure: range 75 (/Table/Max) RANGE_CONSISTENT_STATS_INCORRECT:
        stats: {ContainsEstimates:0 LastUpdateNanos:1746267474019367549 LockAge:1 GCBytesAge:0 LiveBytes:25 LiveCount:1 KeyBytes:15 KeyCount:1 ValBytes:10 ValCount:1 IntentBytes:0 IntentCount:0 LockBytes:95 LockCount:1 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:590 SysCount:8 AbortSpanBytes:0}
        delta (stats-computed): {ContainsEstimates:0 LastUpdateNanos:1746267474019367549 LockAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 LockBytes:0 LockCount:0 RangeKeyCount:0 RangeKeyBytes:0 RangeValCount:0 RangeValBytes:0 SysBytes:-12 SysCount:0 AbortSpanBytes:0}
--- FAIL: TestMergeReplicatesLocks/rhs-lock (35.70s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=41</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeReplicatesLocks.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/kv/kvnemesis/BUILD.bazel
+++ b/pkg/kv/kvnemesis/BUILD.bazel
@@ -86,6 +86,7 @@ go_test(
         "//pkg/kv/kvserver/concurrency/isolation",
         "//pkg/kv/kvserver/kvserverbase",
         "//pkg/kv/kvserver/kvserverpb",
+        "//pkg/kv/kvtestutils",
         "//pkg/roachpb",
         "//pkg/security/securityassets",
         "//pkg/security/securitytest",

--- a/pkg/kv/kvnemesis/kvnemesis_test.go
+++ b/pkg/kv/kvnemesis/kvnemesis_test.go
@@ -22,10 +22,12 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvtestutils"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/testutils/datapathutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
@@ -431,6 +433,7 @@ func testKVNemesisImpl(t *testing.T, cfg kvnemesisTestCfg) {
 	}
 
 	logger := newTBridge(t)
+	defer dumpRaftLogsOnFailure(t, logger.ll.dir, tc.Servers)
 	env := &Env{SQLDBs: sqlDBs, Tracker: tr, L: logger}
 	failures, err := RunNemesis(ctx, rng, env, config, cfg.concurrency, cfg.numSteps, dbs...)
 
@@ -462,4 +465,20 @@ func TestRunReproductionSteps(t *testing.T) {
 	_, _ = db0, ctx
 
 	// Paste a repro as printed by kvnemesis here.
+}
+
+func dumpRaftLogsOnFailure(t *testing.T, dir string, srvs []serverutils.TestServerInterface) {
+	if !t.Failed() {
+		return
+	}
+	d := kvtestutils.RaftLogDumper{Dir: dir}
+	for _, srv := range srvs {
+		require.NoError(t, srv.GetStores().(*kvserver.Stores).VisitStores(func(s *kvserver.Store) error {
+			s.VisitReplicas(func(replica *kvserver.Replica) (wantMore bool) {
+				d.Dump(t, s.LogEngine(), s.StoreID(), replica.RangeID)
+				return true // more
+			})
+			return nil
+		}))
+	}
 }

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -6253,6 +6253,8 @@ func TestMergeReplicatesLocks(t *testing.T) {
 	scope := log.Scope(t)
 	defer scope.Close(t)
 
+	skip.UnderDuress(t, "too slow for testrace")
+
 	// Test Setup:
 	//
 	// txn1: holding lock on key k1

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -6250,7 +6250,8 @@ func TestMergeDropsLocksIfLargerThanMax(t *testing.T) {
 
 func TestMergeReplicatesLocks(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	defer log.Scope(t).Close(t)
+	scope := log.Scope(t)
+	defer scope.Close(t)
 
 	// Test Setup:
 	//
@@ -6288,9 +6289,25 @@ func TestMergeReplicatesLocks(t *testing.T) {
 					Settings: st,
 				},
 			})
+			defer tc.Stopper().Stop(ctx)
+
+			defer func() {
+				if !t.Failed() {
+					return
+				}
+				d := kvtestutils.RaftLogDumper{Dir: scope.GetDirectory()}
+				for _, srv := range tc.Servers {
+					require.NoError(t, srv.GetStores().(*kvserver.Stores).VisitStores(func(s *kvserver.Store) error {
+						s.VisitReplicas(func(replica *kvserver.Replica) (wantMore bool) {
+							d.Dump(t, s.LogEngine(), s.StoreID(), replica.RangeID)
+							return true // more
+						})
+						return nil
+					}))
+				}
+			}()
 
 			sql := tc.ServerConn(0)
-			defer tc.Stopper().Stop(ctx)
 			scratch := tc.ScratchRange(t)
 			mkKey := func(s string) roachpb.Key {
 				prefix := scratch.Clone()

--- a/pkg/kv/kvserver/kvserverpb/proposer_kv.proto
+++ b/pkg/kv/kvserver/kvserverpb/proposer_kv.proto
@@ -299,6 +299,7 @@ message ReplicatedEvalResult {
 // can be distinguished from a zero-length batch, and so structs
 // containing pointers to it can be compared with the == operator.
 message WriteBatch {
+  option (gogoproto.goproto_getters) = true;
   bytes data = 1;
 }
 

--- a/pkg/kv/kvserver/logstore/logstore_test.go
+++ b/pkg/kv/kvserver/logstore/logstore_test.go
@@ -45,7 +45,7 @@ func TestRaftStorageWrites(t *testing.T) {
 		defer batch.Close()
 		prepare(batch)
 		wb := kvserverpb.WriteBatch{Data: batch.Repr()}
-		str, err := print.DecodeWriteBatch(&wb)
+		str, err := print.DecodeWriteBatch(wb.GetData())
 		require.NoError(t, err)
 		require.NoError(t, batch.Commit(true))
 		return str

--- a/pkg/kv/kvserver/print/BUILD.bazel
+++ b/pkg/kv/kvserver/print/BUILD.bazel
@@ -2,7 +2,10 @@ load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
 go_library(
     name = "print",
-    srcs = ["debug_print.go"],
+    srcs = [
+        "debug_print.go",
+        "dump_raft_log.go",
+    ],
     importpath = "github.com/cockroachdb/cockroach/pkg/kv/kvserver/print",
     visibility = ["//visibility:public"],
     deps = [
@@ -18,6 +21,7 @@ go_library(
         "//pkg/util/protoutil",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_cockroachdb_pebble//:pebble",
+        "@com_github_gogo_protobuf//proto",
     ],
 )
 

--- a/pkg/kv/kvserver/print/debug_print.go
+++ b/pkg/kv/kvserver/print/debug_print.go
@@ -232,7 +232,7 @@ func tryIntent(kv storage.MVCCKeyValue) (string, error) {
 	return s, nil
 }
 
-func DecodeWriteBatch(writeBatch *kvserverpb.WriteBatch) (string, error) {
+func DecodeWriteBatch(writeBatch []byte) (string, error) {
 	// Ensure that we always update this function to consider any necessary
 	// updates when a new key kind is introduced. To do this, we assert
 	// pebble.KeyKindDeleteSized is the most recent key kind, ensuring that
@@ -244,7 +244,7 @@ func DecodeWriteBatch(writeBatch *kvserverpb.WriteBatch) (string, error) {
 		return "<nil>\n", nil
 	}
 
-	r, err := storage.NewBatchReader(writeBatch.Data)
+	r, err := storage.NewBatchReader(writeBatch)
 	if err != nil {
 		return "", err
 	}
@@ -381,7 +381,7 @@ func tryRaftLogEntry(kv storage.MVCCKeyValue) (string, error) {
 	e.Data = nil
 	cmd := e.Cmd
 
-	wbStr, err := DecodeWriteBatch(cmd.WriteBatch)
+	wbStr, err := DecodeWriteBatch(cmd.WriteBatch.GetData())
 	if err != nil {
 		wbStr = "failed to decode: " + err.Error() + "\nafter:\n" + wbStr
 	}

--- a/pkg/kv/kvserver/print/debug_print_test.go
+++ b/pkg/kv/kvserver/print/debug_print_test.go
@@ -26,17 +26,20 @@ func TestStringifyWriteBatch(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	wb := &kvserverpb.WriteBatch{}
-	_, err := DecodeWriteBatch(wb)
+	s, err := DecodeWriteBatch(nil)
+	require.NoError(t, err)
+	require.Equal(t, "<nil>\n", s)
+	_, err = DecodeWriteBatch([]byte{})
 	require.ErrorContains(t, err, "batch invalid: too small: 0 bytes")
 
+	wb := &kvserverpb.WriteBatch{}
 	batch := pebble.Batch{}
 	require.NoError(t, batch.Set(storage.EncodeMVCCKey(storage.MVCCKey{
 		Key:       roachpb.Key("/db1"),
 		Timestamp: hlc.Timestamp{WallTime: math.MaxInt64},
 	}), []byte("test value"), nil /* WriteOptions */))
 	wb.Data = batch.Repr()
-	s, err := DecodeWriteBatch(wb)
+	s, err = DecodeWriteBatch(wb.Data)
 	require.NoError(t, err)
 	require.Equal(t, "Put: 9223372036.854775807,0 \"/db1\" (0x2f646231007fffffffffffffff09): \"test value\"\n", s)
 
@@ -46,7 +49,7 @@ func TestStringifyWriteBatch(t *testing.T) {
 	err = batch.SingleDelete(encodedKey, nil)
 	require.NoError(t, err)
 	wb.Data = batch.Repr()
-	s, err = DecodeWriteBatch(wb)
+	s, err = DecodeWriteBatch(wb.Data)
 	require.NoError(t, err)
 	require.Equal(t, "Single Delete: /Local/Lock/Table/56/1/1169/5/3054/0 "+
 		"03623a9318c0384d07a6f22b858594df60 (0x017a6b12c089f704918df70bee8800010003623a9318c0384d07a6f22b858594df6012): \n",
@@ -72,7 +75,7 @@ func TestStringifyWriteBatch(t *testing.T) {
 		nil,
 	))
 	wb.Data = batch.Repr()
-	s, err = DecodeWriteBatch(wb)
+	s, err = DecodeWriteBatch(wb.Data)
 	require.NoError(t, err)
 	require.Equal(t, "Set Range Key: 0.000000001,0 /db{1-2} (0x2f64623100-0x2f64623200): \"\"\n"+
 		"Set Range Key: 0.000000002,0 /db{1-2} (0x2f64623100-0x2f64623200): \"\\x00\\x00\\x00\\x04e\\n\\x02\\b\\x01\"\n",
@@ -92,7 +95,7 @@ func TestStringifyWriteBatch(t *testing.T) {
 		nil,
 	))
 	wb.Data = batch.Repr()
-	s, err = DecodeWriteBatch(wb)
+	s, err = DecodeWriteBatch(wb.Data)
 	require.NoError(t, err)
 	require.Equal(t, "Unset Range Key: 0.000000001,0 /db{1-2} (0x2f64623100-0x2f64623200)\n"+
 		"Unset Range Key: 0.000000002,0 /db{1-2} (0x2f64623100-0x2f64623200)\n",
@@ -105,7 +108,7 @@ func TestStringifyWriteBatch(t *testing.T) {
 		nil,
 	))
 	wb.Data = batch.Repr()
-	s, err = DecodeWriteBatch(wb)
+	s, err = DecodeWriteBatch(wb.Data)
 	require.NoError(t, err)
 	require.Equal(t, "Delete Range Keys: /db{1-2} (0x2f64623100-0x2f64623200)\n", s)
 }

--- a/pkg/kv/kvserver/print/dump_raft_log.go
+++ b/pkg/kv/kvserver/print/dump_raft_log.go
@@ -31,11 +31,11 @@ func DumpRaftLog(w io.Writer, reader storage.Reader, rangeID roachpb.RangeID) er
 	return raftlog.Visit(ctx, reader, rangeID, 0, math.MaxUint64, func(ent raftpb.Entry) error {
 		put("****** index %d ******", ent.Index)
 		e, err := raftlog.NewEntry(ent)
-		defer e.Release()
 		if err != nil {
 			put("%s", err)
 			return nil // continue
 		}
+		defer e.Release()
 		wb := e.Cmd.WriteBatch
 		e.Cmd.WriteBatch = nil
 
@@ -49,7 +49,7 @@ func DumpRaftLog(w io.Writer, reader storage.Reader, rangeID roachpb.RangeID) er
 		s, err := DecodeWriteBatch(wb.GetData())
 		if err != nil {
 			put("%s", err)
-			// continue
+			return nil // continue
 		}
 		put("%s", strings.TrimSpace(s))
 		return nil

--- a/pkg/kv/kvserver/print/dump_raft_log.go
+++ b/pkg/kv/kvserver/print/dump_raft_log.go
@@ -1,0 +1,57 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package print
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/raftlog"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/gogo/protobuf/proto"
+)
+
+// DumpRaftLog writes a debugging text representation of the raft log into the
+// provided writer.
+func DumpRaftLog(w io.Writer, reader storage.Reader, rangeID roachpb.RangeID) error {
+	ctx := context.Background()
+	put := func(format string, args ...interface{}) {
+		_, _ = fmt.Fprintf(w, format, args...)
+		_, _ = fmt.Fprintln(w)
+	}
+
+	return raftlog.Visit(ctx, reader, rangeID, 0, math.MaxUint64, func(ent raftpb.Entry) error {
+		put("****** index %d ******", ent.Index)
+		e, err := raftlog.NewEntry(ent)
+		defer e.Release()
+		if err != nil {
+			put("%s", err)
+			return nil // continue
+		}
+		wb := e.Cmd.WriteBatch
+		e.Cmd.WriteBatch = nil
+
+		// NB: I spent way too much time trying to find a way to output the
+		// struct recursively but while omitting zero values. This is the
+		// closest I got without putting in significant amounts of work (like
+		// adjusting dozens of gogoproto moretags or writing custom
+		// formatters).
+		put("RaftCommand: %+v", proto.CompactTextString(&e.Cmd))
+
+		s, err := DecodeWriteBatch(wb.GetData())
+		if err != nil {
+			put("%s", err)
+			// continue
+		}
+		put("%s", strings.TrimSpace(s))
+		return nil
+	})
+}

--- a/pkg/kv/kvtestutils/BUILD.bazel
+++ b/pkg/kv/kvtestutils/BUILD.bazel
@@ -7,6 +7,7 @@ go_library(
     # testonly = 1,
     srcs = [
         "consistency.go",
+        "dump_raft_logs.go",
         "test_utils.go",
     ],
     importpath = "github.com/cockroachdb/cockroach/pkg/kv/kvtestutils",
@@ -14,8 +15,10 @@ go_library(
     deps = [
         "//pkg/kv/kvbase",
         "//pkg/kv/kvpb",
+        "//pkg/kv/kvserver/print",
         "//pkg/roachpb",
         "//pkg/sql/pgwire/pgerror",
+        "//pkg/storage",
         "//pkg/util/tracing",
         "//pkg/util/tracing/tracingpb",
         "@com_github_cockroachdb_errors//:errors",

--- a/pkg/kv/kvtestutils/dump_raft_logs.go
+++ b/pkg/kv/kvtestutils/dump_raft_logs.go
@@ -1,0 +1,55 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package kvtestutils
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/print"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+)
+
+type RaftLogDumper struct {
+	Dir    string
+	logged bool
+}
+
+func (d *RaftLogDumper) Dump(
+	t interface {
+		Logf(format string, args ...interface{})
+	},
+	reader storage.Reader,
+	storeID roachpb.StoreID,
+	rangeID roachpb.RangeID,
+) {
+	if !d.logged {
+		t.Logf("dumping raft logs to %s", d.Dir)
+		d.logged = true
+	}
+	if err := d.dumpOne(reader, storeID, rangeID); err != nil {
+		t.Logf("error dumping s%dr%d: %s", storeID, rangeID, err)
+	}
+}
+
+func (d *RaftLogDumper) dumpOne(
+	reader storage.Reader, storeID roachpb.StoreID, rangeID roachpb.RangeID,
+) error {
+	if err := os.MkdirAll(d.Dir, 0755); err != nil {
+		return err
+	}
+	path := filepath.Join(d.Dir, "raftlog_s"+storeID.String()+"r"+rangeID.String()+".txt")
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	if err := print.DumpRaftLog(f, reader, rangeID); err != nil {
+		_, _ = fmt.Fprintf(f, "error on %s: %s", rangeID, err)
+	}
+	return f.Close()
+}

